File: task4/main.cpp
#include <iostream>
#include <vector>

using namespace std;

// Проверка корректности одного голоса:
// каждый кандидат должен встретиться ровно один раз
bool checkVote(const vector<int>& vote, int n) {
    if ((int)vote.size() != n) {
        return false;
    }

    vector<int> used(n + 1, 0);

    for (int candidate : vote) {
        if (candidate < 1 || candidate > n) {
            return false;
        }

        if (used[candidate]) {
            return false;
        }

        used[candidate] = 1;
    }

    return true;
}

// Метод Борда
// Возвращает номер победителя.
// Если победителя нет из-за равенства баллов, возвращает -1.
int findBordaWinner(const vector<vector<int>>& votes, int n) {
    vector<int> points(n + 1, 0);

    for (const vector<int>& vote : votes) {
        for (int place = 0; place < n; ++place) {
            int candidate = vote[place];
            points[candidate] += n - 1 - place;
        }
    }

    cout << "\nМетод Борда:\n";

    for (int candidate = 1; candidate <= n; ++candidate) {
        cout << "Кандидат " << candidate
             << ": " << points[candidate] << " баллов\n";
    }

    int maxPoints = points[1];

    for (int candidate = 2; candidate <= n; ++candidate) {
        if (points[candidate] > maxPoints) {
            maxPoints = points[candidate];
        }
    }

    vector<int> winners;

    for (int candidate = 1; candidate <= n; ++candidate) {
        if (points[candidate] == maxPoints) {
            winners.push_back(candidate);
        }
    }

    if (winners.size() > 1) {
        cout << "Победитель по методу Борда отсутствует: равенство баллов между кандидатами ";

        for (int candidate : winners) {
            cout << candidate << " ";
        }

        cout << endl;
        return -1;
    }

    return winners[0];
}

// Проверяет, стоит ли candidateA выше candidateB в голосе одного избирателя
bool isHigher(const vector<int>& vote, int candidateA, int candidateB) {
    for (int candidate : vote) {
        if (candidate == candidateA) {
            return true;
        }

        if (candidate == candidateB) {
            return false;
        }
    }

    return false;
}

// Метод Кондорсе
// Возвращает номер победителя.
// Если победителя нет, возвращает -1.
int findCondorcetWinner(const vector<vector<int>>& votes, int n) {
    for (int candidate = 1; candidate <= n; ++candidate) {
        bool winsAll = true;

        for (int opponent = 1; opponent <= n; ++opponent) {
            if (candidate == opponent) {
                continue;
            }

            int candidateWins = 0;
            int opponentWins = 0;

            for (const vector<int>& vote : votes) {
                if (isHigher(vote, candidate, opponent)) {
                    ++candidateWins;
                } else {
                    ++opponentWins;
                }
            }

            if (candidateWins <= opponentWins) {
                winsAll = false;
                break;
            }
        }

        if (winsAll) {
            return candidate;
        }
    }

    return -1;
}

int main() {
    int n, k;

    cout << "Введите количество кандидатов: ";
    cin >> n;

    cout << "Введите количество избирателей: ";
    cin >> k;

    if (n < 2 || k < 1) {
        cout << "Ошибка: кандидатов должно быть не меньше 2, избирателей не меньше 1.\n";
        return 1;
    }

    vector<vector<int>> votes(k, vector<int>(n));

    cout << "\nВведите предпочтения каждого избирателя.\n";
    cout << "Каждая строка - номера кандидатов от лучшего к худшему.\n";

    for (int i = 0; i < k; ++i) {
        cout << "Избиратель " << i + 1 << ": ";

        for (int j = 0; j < n; ++j) {
            cin >> votes[i][j];
        }

        if (!checkVote(votes[i], n)) {
            cout << "Ошибка: некорректное ранжирование у избирателя "
                 << i + 1 << ".\n";
            return 1;
        }
    }

    int bordaWinner = findBordaWinner(votes, n);

    if (bordaWinner != -1) {
        cout << "Победитель по методу Борда: кандидат "
             << bordaWinner << endl;
    }

    cout << "\nМетод Кондорсе:\n";

    int condorcetWinner = findCondorcetWinner(votes, n);

    if (condorcetWinner == -1) {
        cout << "Победитель Кондорсе отсутствует.\n";
    } else {
        cout << "Победитель по методу Кондорсе: кандидат "
             << condorcetWinner << endl;
    }

    return 0;
}

File: task4/main.rs
use std::io;

// Чтение строки чисел
fn read_numbers() -> Vec<usize> {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();

    input
        .split_whitespace()
        .map(|x| x.parse::<usize>().unwrap())
        .collect()
}

// Проверка корректности голоса:
// каждый кандидат должен встретиться ровно один раз
fn check_vote(vote: &Vec<usize>, n: usize) -> bool {
    if vote.len() != n {
        return false;
    }

    let mut used = vec![false; n + 1];

    for &candidate in vote {
        if candidate < 1 || candidate > n {
            return false;
        }

        if used[candidate] {
            return false;
        }

        used[candidate] = true;
    }

    true
}

// Метод Борда.
// Возвращает Some(номер кандидата), если победитель один.
// Возвращает None, если победителя нет из-за равенства баллов.
fn find_borda_winner(votes: &Vec<Vec<usize>>, n: usize) -> Option<usize> {
    let mut points = vec![0usize; n + 1];

    for vote in votes {
        for place in 0..n {
            let candidate = vote[place];
            points[candidate] += n - 1 - place;
        }
    }

    println!("\nМетод Борда:");

    for candidate in 1..=n {
        println!("Кандидат {}: {} баллов", candidate, points[candidate]);
    }

    let mut max_points = points[1];

    for candidate in 2..=n {
        if points[candidate] > max_points {
            max_points = points[candidate];
        }
    }

    let mut winners: Vec<usize> = Vec::new();

    for candidate in 1..=n {
        if points[candidate] == max_points {
            winners.push(candidate);
        }
    }

    if winners.len() > 1 {
        print!("Победитель по методу Борда отсутствует: равенство баллов между кандидатами ");

        for candidate in winners {
            print!("{} ", candidate);
        }

        println!();
        return None;
    }

    Some(winners[0])
}

// Проверяет, стоит ли candidate_a выше candidate_b в голосе одного избирателя
fn is_higher(vote: &Vec<usize>, candidate_a: usize, candidate_b: usize) -> bool {
    for &candidate in vote {
        if candidate == candidate_a {
            return true;
        }

        if candidate == candidate_b {
            return false;
        }
    }

    false
}

// Метод Кондорсе.
// Возвращает Some(номер кандидата), если победитель есть.
// Возвращает None, если победителя нет.
fn find_condorcet_winner(votes: &Vec<Vec<usize>>, n: usize) -> Option<usize> {
    for candidate in 1..=n {
        let mut wins_all = true;

        for opponent in 1..=n {
            if candidate == opponent {
                continue;
            }

            let mut candidate_wins = 0usize;
            let mut opponent_wins = 0usize;

            for vote in votes {
                if is_higher(vote, candidate, opponent) {
                    candidate_wins += 1;
                } else {
                    opponent_wins += 1;
                }
            }

            if candidate_wins <= opponent_wins {
                wins_all = false;
                break;
            }
        }

        if wins_all {
            return Some(candidate);
        }
    }

    None
}

fn main() {
    println!("Введите количество кандидатов:");
    let n = read_numbers()[0];

    println!("Введите количество избирателей:");
    let k = read_numbers()[0];

    if n < 2 || k < 1 {
        println!("Ошибка: кандидатов должно быть не меньше 2, избирателей не меньше 1.");
        return;
    }

    let mut votes: Vec<Vec<usize>> = Vec::new();

    println!("\nВведите предпочтения каждого избирателя.");
    println!("Каждая строка - номера кандидатов от лучшего к худшему.");

    for i in 0..k {
        println!("Избиратель {}:", i + 1);

        let vote = read_numbers();

        if !check_vote(&vote, n) {
            println!("Ошибка: некорректное ранжирование у избирателя {}.", i + 1);
            return;
        }

        votes.push(vote);
    }

    match find_borda_winner(&votes, n) {
        Some(winner) => {
            println!("Победитель по методу Борда: кандидат {}", winner);
        }
        None => {}
    }

    println!("\nМетод Кондорсе:");

    match find_condorcet_winner(&votes, n) {
        Some(winner) => {
            println!("Победитель по методу Кондорсе: кандидат {}", winner);
        }
        None => {
            println!("Победитель Кондорсе отсутствует.");
        }
    }
}
